File: 216-combination-sum-iii/combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        result = []

        def backtrack(start: int, combination: List[int], remaining: int):
            if len(combination) == k and remaining == 0:
                result.append(combination[:])
                return
            
            if len(combination) > k or remaining < 0:
                return

            for i in range(start, 10):
                combination.append(i)
                backtrack(i + 1, combination, remaining - i)
                combination.pop()
            
        backtrack(1, [], n)
        return result

File: 216-combination-sum-iii/combination-sum-iii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> result;
        vector<int> combination;

        function<void(int, int)> backtrack = [&](int start, int remaining) {
            if (combination.size() == k && remaining == 0) {
                result.push_back(combination);
                return;
            }

            if (combination.size() > k || remaining == 0) {
                return;
            }
        
            for (int i = start; i <= 9; ++i) {
                combination.push_back(i);          
                backtrack(i + 1, remaining - i);   
                combination.pop_back();           
            }  
        };

        backtrack(1, n);
        return result;
    }
};